pkg/internal/templates.go

```diff
@@ -0,0 +1,225 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package internal
+
+const wrappedInterfacesTempl = `
+
+//go:build !ignore_autogenerated
+// +build !ignore_autogenerated
+
+// Code auto-generated. DO NOT EDIT.
+
+package {{.ClientsetName}}
+
+import (
+	"fmt"
+	
+	kcp "github.com/kcp-dev/apimachinery/pkg/client"
+	"github.com/kcp-dev/apimachinery/pkg/logicalcluster"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"{{.ClientsetAPIPath}}"
+
+	{{$clientPath := .ClientsetAPIPath}}
+	{{$pkg := .TypedPkgPath}}
+	{{ range .APIs }}
+	{{.PkgName}}{{.Version}} "{{$clientPath}}/typed/{{.PkgName}}/{{.Version}}"
+	{{.PkgName}}{{.Version}}client "{{$pkg}}/typed/{{.PkgName}}/{{.Version}}"
+
+	{{ end }}
+)
+
+func NewForConfig(config *rest.Config) (*ClusterClient, error) {
+	client, err := rest.HTTPClientFor(config)
+	if err != nil {
+		return nil, fmt.Errorf("error creating HTTP client: %w", err)
+	}
+
+	clusterRoundTripper := kcp.NewClusterRoundTripper(client.Transport)
+	client.Transport = clusterRoundTripper
+
+	delegate, err := {{.InterfaceName}}.NewForConfigAndClient(config, client)
+	if err != nil {
+		return nil, fmt.Errorf("error creating delegate clientset: %w", err)
+	}
+
+	return &ClusterClient{
+		delegate: delegate,
+	}, nil
+}
+
+type ClusterClient struct {
+	delegate {{.InterfaceName}}.Interface
+}
+
+func (c *ClusterClient) Cluster(cluster logicalcluster.LogicalCluster) {{.InterfaceName}}.Interface {
+	return &wrappedInterface{
+		cluster:  cluster,
+		delegate: c.delegate,
+	}
+}
+
+type wrappedInterface struct {
+	cluster  logicalcluster.LogicalCluster
+	delegate {{.InterfaceName}}.Interface
+}
+
+func (w *wrappedInterface) Discovery() discovery.DiscoveryInterface {
+	return w.delegate.Discovery()
+}
+
+{{ range .APIs }}
+func (w *wrappedInterface) {{.PkgNameUpperFirst}}{{.VersionUpperFirst}}() {{.PkgName}}{{.Version}}.{{.PkgNameUpperFirst}}{{.VersionUpperFirst}}Interface {
+	return &{{.PkgName}}{{.Version}}client.Wrapped{{.PkgNameUpperFirst}}{{.VersionUpperFirst}}{
+		Cluster:  w.cluster,
+		Delegate: w.delegate.{{.PkgNameUpperFirst}}{{.VersionUpperFirst}}(),
+	}
+}
+{{ end }}
+
+`
+
+const commonTempl = `
+
+//go:build !ignore_autogenerated
+// +build !ignore_autogenerated
+// Code auto-generated. DO NOT EDIT.
+
+package {{.Version}}
+
+import (
+	"context"
+	"fmt"
+	{{.Name}}api{{.Version}} "{{.APIPath}}"
+	{{.Name}}{{.Version}} "{{.ClientPath}}/typed/{{.Name}}/{{.Version}}"
+
+	kcp "github.com/kcp-dev/apimachinery/pkg/client"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	"github.com/kcp-dev/apimachinery/pkg/logicalcluster"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+type Wrapped{{.NameUpperFirst}}{{.VersionUpperFirst}} struct {
+	Cluster  logicalcluster.LogicalCluster
+	Delegate {{.Name}}{{.Version}}.{{.NameUpperFirst}}{{.VersionUpperFirst}}Interface
```

Comment:
That's because the wrapper for typed apis is defined separately, and is needed here (ex): https://github.com/varshaprasad96/custom-crd-operator/blob/7f26b5e54c536812f6ef7adafd3214971f3f032f/pkg/client/clusterclient/clientset.go#L76. If we have all of these in the same package then that would solve the problem (I agree on not wanting it to be exported, `Delegate` field especially). But not sure if that's what we want since it would be clumsy to have all of the wrappers in the same file. The other option I could think of is specify non-keyed fields, but again linter complains (again not sure if we want that).

Reply:
Instead of `return &examplev1alpha1Client.WrappedExampleV1alpha1{`, you could do `return new examplev1alpha1Client.WrappedExampleV1alpha1(w.cluster, w.delegate.ExampleV1alpha1())`.

Also please make sure the package import aliases are all lowercase

Reply:
Sorry for the back and forth on this. Did you mean this: https://github.com/kcp-dev/code-generator/pull/1/files#diff-9bbc92b4a909f3d2fd2c0e5eecfdd8060d332ac794cb121fed560c97d9e1374cR43 and call it here: https://github.com/kcp-dev/code-generator/pull/1/files#diff-3d08ab7ce03a59265189e044478e0fcd57c9d862dbbcaa7cc6278a547eebb6bfR76

Reply:
In `custom-crd-operator/pkg/client/clusterclient/clientset.go` you have

```go
func (w *wrappedInterface) ExampleV1alpha1() examplev1alpha1.ExampleV1alpha1Interface {
	return &examplev1alpha1Client.WrappedExampleV1alpha1{
		Cluster:  w.cluster,
		Delegate: w.delegate.ExampleV1alpha1(),
	}
}
```

You said that's why you need Cluster and Delegate exported. I'm suggesting you instead do

```go
func (w *wrappedInterface) ExampleV1alpha1() examplev1alpha1.ExampleV1alpha1Interface {
	return examplev1alpha1Client.NewWrappedExampleV1alpha1(w.cluster, w.delegate.ExampleV1alpha1())
}
```

or whatever makes the most sense so you're calling a constructor instead of trying to fill out the fields in the struct directly. Then you don't need to export them.

Reply:
cool, thanks. updated and did it here: https://github.com/kcp-dev/code-generator/blob/727e3e4ae4bafa0259eff108ad63278a33b17d2c/testdata/pkg/clusterclient/clientset.go#L77